### PR TITLE
GDB-4761 Prevent back-end actions and show error on incomplete triples

### PR DESCRIPTION
**Pull request by teodossidossev**
Validation error are present an all back-end requests when we have an incomplete triple

### CODE REVIEW

File: src/app/main/mapper/iteration/iteration.component.ts [hunk 1]
```diff
@@ -162,7 +169,8 @@ export class IterationComponent extends OnDestroyMixin implements OnInit, AfterV
       this.mapping.getSubjectMappings().length &&
       this.isPreviewOn &&
       this.isComplete(this.mapping) &&
```
Comment (svilenvelikov) [inline src/app/main/mapper/iteration/iteration.component.ts:171]: Holly shit... Can this whole expression be simplified somehow or have some local variables with meaningful names? This is just unbearably ugly.

File: src/app/main/mapper/cell/cell.component.ts [hunk 2]
```diff
@@ -271,12 +271,19 @@ export class CellComponent extends OnDestroyMixin implements OnInit {
...
   public onDropEvent(event) {
     this.onDrop.emit(event);
   }
+
+  public isIncomplete() {
+    return !this.isEmptyTriple() && this.isEmpty();
+  }
+
+  private isEmptyTriple() {
+    return !this.triple.getSubject() && !this.triple.getPredicate() && !this.triple.getObject();
```
Comment (svilenvelikov) [inline src/app/main/mapper/cell/cell.component.ts:287]: Can we have such method in the Triple itself? I think it'd suit better there.

File: src/app/main/mapper/mapper.component.ts [hunk 3]
```diff
@@ -64,32 +68,53 @@ export class MapperComponent extends OnDestroyMixin implements OnInit {
...
+          } else {
+            this.showErrorMessage();
+          }
         });
 
     this.messageService.read(ChannelName.GetRDF)
         .pipe(untilComponentDestroyed(this))
         .subscribe(() => {
-          this.onGetRDF();
+          if (this.modelManagementService.isCompleteMapping(this.mapping)) {
```
Comment (svilenvelikov) [inline src/app/main/mapper/mapper.component.ts:85]: As these are duplicated 4 times, I wonder, can we have a method that accepts a method reference here or in the modelManagementService that do this check and error reporting?

File: src/app/services/model-management.service.ts [hunk 4]
```diff
@@ -420,4 +420,29 @@ export class ModelManagementService {
...
+      this.checkMappingValidity(mapping);
+      return true;
+    } catch (e) {
+      return false;
+    }
+  }
+
+  private checkSubjectValidity(mapping: MappingDefinitionImpl) {
+    mapping.getSubjectMappings().forEach((subject) => {
+      const count = subject.getPropertyMappings().length + subject.getTypeMappings().length;
```
Comment (svilenvelikov) [inline src/app/services/model-management.service.ts:436]: Is there a chance that propertyMappings or typeMappings are undefined?

Comment (teodossidossev) [reply src/app/services/model-management.service.ts:436]: at the very beginning of a subject mapping - yes

File: src/app/main/mapper/cell/cell.component.ts (hunk 2 above)
Comment (teodossidossev) [reply src/app/main/mapper/cell/cell.component.ts:287]: +1

File: src/app/main/mapper/iteration/iteration.component.ts (hunk 1 above)
Comment (teodossidossev) [reply src/app/main/mapper/iteration/iteration.component.ts:171]: sure

File: src/app/main/mapper/mapper.component.ts (hunk 3 above)
Comment (teodossidossev) [reply src/app/main/mapper/mapper.component.ts]: +1

File: src/app/main/mapper/mapper.component.ts [hunk 5]
```diff
@@ -64,32 +68,49 @@ export class MapperComponent extends OnDestroyMixin implements OnInit {
...
+    if (this.modelManagementService.isCompleteMapping(this.mapping)) {
+      command();
+    } else {
+      this.showErrorMessage();
+    }
+  }
+
+  private showErrorMessage() {
+    this.notificationService.error(this.translateService.instant('MESSAGES.INCOMPLETE_MAPPING_ERROR'));
+    this.messageService.publish(ChannelName.ProgressCancelled);
```
Comment (svilenvelikov) [inline src/app/main/mapper/mapper.component.ts:111]: I'd prefer this event to be fired in the method client and not here. The purpose of the method is to rise notification and not to fire this event

Comment (teodossidossev) [reply src/app/main/mapper/mapper.component.ts:111]: ok

File: src/app/services/model-management.service.ts (hunk 4 above)
Comment (svilenvelikov) [reply src/app/services/model-management.service.ts:436]: So, can this blow then?

Comment (teodossidossev) [reply src/app/services/model-management.service.ts:436]: right

File: src/app/main/mapper/iteration/iteration.component.ts (hunk 1 above)
Comment (svilenvelikov) [reply src/app/main/mapper/iteration/iteration.component.ts:171]: did you forget to push?

**Pull request merged**